### PR TITLE
Kafka 2.4 (for review)

### PR DESCRIPTION
**Pull request by dkaukov**
1) Upgrade kafka
2) Fixed tests
3) otj-kafka-parent - which is replacement for org.springframework.kafka:spring-kafka-test

### CODE REVIEW

**Comment by dkaukov** [comment]
@mikebell90 
This one properly upgrades to the 2.4 and creates `com.opentable.components:otj-spring-kafka-test` which should be used instead of the `org.springframework.kafka:spring-kafka-test` till we upgrade spring to 5.2. Then, we can just drop it and relocate.

File: otj-kafka-spring/pom.xml [hunk 1]
```diff
@@ -93,10 +93,15 @@
...
+        <dependency>
+            <groupId>com.opentable.components</groupId>
+            <artifactId>otj-spring-kafka-test</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>org.springframework.kafka</groupId>
             <artifactId>spring-kafka-test</artifactId>
-            <scope>test</scope>
+            <scope>provided</scope>
```
Comment (mikebell90) [inline otj-kafka-spring/pom.xml]: Why did this change?

File: otj-kafka-spring/src/test/java/com/opentable/kafka/spring/features/BatchListenerTest.java [hunk 2]
```diff
@@ -102,7 +102,7 @@
             return new SettableListenableFuture<>();
         }
 
-        @KafkaListener(id = "foo", topics = "topic-1", containerFactory = "kafkaListenerContainerFactory", concurrency = "1")
+        @KafkaListener(id = "foo", topics = "topic-1", containerFactory = "kafkaListenerContainerFactory", concurrency = "1", autoStartup = "false")
```
Comment (mikebell90) [inline otj-kafka-spring/src/test/java/com/opentable/kafka/spring/features/BatchListenerTest.java]: what does this do?

File: otj-kafka/src/test/java/com/opentable/kafka/session/BaseSessionTest.java [hunk 3]
```diff
@@ -63,7 +63,7 @@ public void performTestSimpleWithStategy() throws ExecutionException, Interrupte
 
         // Has consumed a total of N messages
         assertThat(eventTypes.stream().filter(t -> t == ConsumedEvent.EventType.POLL).count()).isGreaterThanOrEqualTo(totalMessages);
-        moreAssertionsForTestResult(withSleep, simpleTestResult, (long) totalMessages, withSleep ? 1L : 0L);
+        moreAssertionsForTestResult(withSleep, simpleTestResult, (long) totalMessages, withSleep ? 2L : 1L);
```
Comment (mikebell90) [inline otj-kafka/src/test/java/com/opentable/kafka/session/BaseSessionTest.java]: Ok, so here I really need a clear explanation of why these changed, and why these fixes were needed.

File: otj-kafka/src/test/java/com/opentable/kafka/session/BaseSessionTest.java [hunk 4]
```diff
@@ -103,7 +103,7 @@ public void performComplexTestWithNoStrategy() throws ExecutionException, Interr
         Set<UUID> consumerIds = messageEvents.stream().map(ConsumedEvent::getConsumerId).collect(Collectors.toSet());
         assertThat(consumerIds).hasSize(3);
         assertThat(eventTypeList).containsOnly(ConsumedEvent.EventType.SUBSCRIBED, ConsumedEvent.EventType.MESSAGE, ConsumedEvent.EventType.CLOSING, ConsumedEvent.EventType.POLL);
-        moreAssertionsForTestResult(false, complexTestResult,100L,0L);
+        moreAssertionsForTestResult(false, complexTestResult,100L,3L);
```
Comment (mikebell90) [inline otj-kafka/src/test/java/com/opentable/kafka/session/BaseSessionTest.java]: same quesrtion here,

File: otj-spring-kafka-test/pom.xml [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    <groupId>com.opentable.components</groupId>
+    <version>3.2.3-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>otj-spring-kafka-test</artifactId>
+  <version>3.2.3-SNAPSHOT</version>
+
+  <properties>
+    <basepom.check.skip-pmd>true</basepom.check.skip-pmd>
```
Comment (mikebell90) [inline otj-spring-kafka-test/pom.xml]: Why skip dependencies and pmd?

File: pom.xml [hunk 6]
```diff
@@ -38,8 +38,19 @@
         <!-- We can't go to Curator 3 or 4 without bumping to Zookeeper 3.5. Since Kafka is still shipping with 3.4, holding off -->
         <dep.curator.version>2.13.0</dep.curator.version>
 
+
+        <dep.kafka.version>2.4.0</dep.kafka.version>
+        <!-- required for kafka 2.4 -->
+        <dep.zookeeper.version>3.5.6</dep.zookeeper.version>
+        <!-- not strictly required, but might as well use an up to date -->
+       <!-- <dep.curator.version>4.2.0</dep.curator.version>-->
```
Comment (mikebell90) [inline pom.xml]: what about curator?

**Comment by mikebell90** [comment]
So this looks interesting but we should chat
If I read correctly, you upgrade kafka to 2.4.0, but leave spring kafka at 2.2 (which I worry about) and try to graft on these classes to compensate? Sorry I'm confused here, though I agree that since you are the biggest user of spring-kafka it wold unblock quite a bit.
Also dont understand the test fixes.

**Comment by mikebell90** [comment]
Various bits stolen or ignored in #47

**Pull request closed without merging**